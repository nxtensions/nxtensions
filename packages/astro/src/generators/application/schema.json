{
  "$schema": "http://json-schema.org/schema",
  "$id": "NxtensionsAstroApplicationGenerator",
  "title": "Astro Application Generator",
  "description": "Generates an Astro application.",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the application.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?",
      "pattern": "^[a-zA-Z].*$"
    },
    "tags": {
      "type": "string",
      "description": "Comma separated list of tags to be applied to the project (used for linting).",
      "alias": "t"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed.",
      "alias": "d"
    },
    "integrations": {
      "type": "array",
      "description": "Frameworks and addons to enable in the application.",
      "alias": "i",
      "default": [],
      "x-prompt": {
        "message": "What frameworks and/or addons would you like to enable?",
        "type": "list",
        "multiselect": true,
        "items": [
          {
            "label": "Lit",
            "value": "lit"
          },
          {
            "label": "Preact",
            "value": "preact"
          },
          {
            "label": "React",
            "value": "react"
          },
          {
            "label": "Solid",
            "value": "solid-js"
          },
          {
            "label": "Svelte",
            "value": "svelte"
          },
          {
            "label": "Vue",
            "value": "vue"
          },
          {
            "label": "Partytown",
            "value": "partytown"
          },
          {
            "label": "Sitemap",
            "value": "sitemap"
          },
          {
            "label": "Tailwind CSS",
            "value": "tailwind"
          },
          {
            "label": "Turbolinks",
            "value": "turbolinks"
          }
        ]
      }
    },
    "standaloneConfig": {
      "description": "Split the project configuration into `<projectRoot>/project.json` rather than including it inside `workspace.json`.",
      "type": "boolean"
    },
    "addCypressTests": {
      "description": "Generate an E2E project with Cypress.",
      "type": "boolean",
      "default": true
    }
  },
  "additionalProperties": false,
  "required": ["name"]
}
