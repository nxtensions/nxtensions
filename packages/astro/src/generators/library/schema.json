{
  "$schema": "http://json-schema.org/schema",
  "$id": "NxtensionsAstroLibraryGenerator",
  "title": "Astro Library Generator",
  "description": "Generates an Astro library.",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the library.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?",
      "pattern": "^[a-zA-Z].*$"
    },
    "tags": {
      "type": "string",
      "description": "Comma separated list of tags to be applied to the project (used for linting).",
      "alias": "t"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed.",
      "alias": "d"
    },
    "projectNameAndRootFormat": {
      "description": "Whether to generate the project name and root directory as provided (`as-provided`) or generate them composing their values and taking the configured layout into account (`derived`).",
      "type": "string",
      "enum": ["as-provided", "derived"]
    },
    "publishable": {
      "type": "boolean",
      "description": "Whether the library is intended to be distributed.",
      "default": false
    },
    "importPath": {
      "type": "string",
      "description": "The path used to import the library, like `@myorg/my-awesome-lib`. Must be a valid npm name and must be provided when creating a publishable library."
    },
    "standaloneConfig": {
      "description": "Split the project configuration into `<projectRoot>/project.json` rather than including it inside `workspace.json`.",
      "type": "boolean",
      "x-deprecated": "This option is no longer used and will be removed in v19."
    }
  },
  "additionalProperties": false,
  "required": ["name"]
}
