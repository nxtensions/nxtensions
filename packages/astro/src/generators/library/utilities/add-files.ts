import {
  generateFiles,
  joinPathFragments,
  names,
  offsetFromRoot,
  type Tree,
} from '@nx/devkit';
import type { NormalizedGeneratorOptions } from '../schema';

export function addFiles(tree: Tree, options: NormalizedGeneratorOptions) {
  generateFiles(
    tree,
    joinPathFragments(__dirname, '..', 'files'),
    options.projectRoot,
    {
      ...options,
      ...names(options.projectName),
      offsetFromRoot: offsetFromRoot(options.projectRoot),
      tmpl: '',
    }
  );

  if (!options.publishable) {
    tree.delete(joinPathFragments(options.projectRoot, 'package.json'));
  }
}
