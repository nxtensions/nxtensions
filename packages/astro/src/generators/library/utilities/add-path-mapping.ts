import type { Tree } from '@nrwl/devkit';
import { joinPathFragments, updateJson } from '@nrwl/devkit';
import type { NormalizedGeneratorOptions } from '../schema';

export function addPathMapping(
  tree: Tree,
  options: NormalizedGeneratorOptions
): void {
  const rootTsConfigPath = getRootTsConfigPath(tree);

  if (!rootTsConfigPath) {
    return;
  }

  updateJson(tree, rootTsConfigPath, (json) => {
    const c = json.compilerOptions;
    c.paths = c.paths ?? {};
    c.paths[options.importPath] = [
      joinPathFragments(options.projectRoot, 'src/index.js'),
    ];

    return json;
  });
}

function getRootTsConfigPath(tree: Tree): string | undefined {
  const possiblePaths = ['tsconfig.base.json', 'tsconfig.json'];

  for (const path of possiblePaths) {
    if (tree.exists(path)) {
      return path;
    }
  }

  return undefined;
}
