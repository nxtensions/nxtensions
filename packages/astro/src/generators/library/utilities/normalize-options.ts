import type { Tree } from '@nx/devkit';
import { determineProjectNameAndRootOptions } from '@nx/devkit/src/generators/project-name-and-root-utils';
import type { GeneratorOptions, NormalizedGeneratorOptions } from '../schema';

export async function normalizeOptions(
  tree: Tree,
  options: GeneratorOptions
): Promise<NormalizedGeneratorOptions> {
  const { importPath, projectName, projectRoot } =
    await determineProjectNameAndRootOptions(tree, {
      name: options.name,
      projectType: 'library',
      directory: options.directory,
      importPath: options.importPath,
    });

  const tags = options.tags ? options.tags.split(',').map((s) => s.trim()) : [];

  return {
    ...options,
    importPath,
    projectName,
    projectRoot,
    publishable: options.publishable ?? false,
    tags,
  };
}
