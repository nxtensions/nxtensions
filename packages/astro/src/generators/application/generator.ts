import {
  ensurePackage,
  formatFiles,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';
import { addDependenciesToPackageJson } from '../../utilities/package-json';
import { initGenerator } from '../init/generator';
import {
  astroCheckVersion,
  astroVersion,
  getInstalledNxVersion,
  typescriptVersion,
} from '../utilities/versions';
import type { GeneratorOptions } from './schema';
import {
  addFiles,
  addIntegrationsPackages,
  addProject,
  configureTailwindIntegration,
  normalizeOptions,
  setDefaultProject,
  setupE2ETests,
} from './utilities';

export async function applicationGenerator(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback | void> {
  return await applicationGeneratorInternal(tree, {
    projectNameAndRootFormat: 'derived',
    ...rawOptions,
  });
}

export async function applicationGeneratorInternal(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback | void> {
  const options = await normalizeOptions(tree, rawOptions);

  const tasks: GeneratorCallback[] = [];
  const { initGenerator: jsInitGenerator } = ensurePackage<
    typeof import('@nx/js')
  >('@nx/js', getInstalledNxVersion(tree));
  const jsTask = await jsInitGenerator(tree, { skipFormat: true });
  tasks.push(jsTask);

  await initGenerator(tree, {});

  const depsTask = addDependenciesToPackageJson(
    tree,
    {},
    {
      astro: astroVersion,
      '@astrojs/check': astroCheckVersion,
      typescript: typescriptVersion,
    }
  );
  tasks.push(depsTask);

  addProject(tree, options);
  setDefaultProject(tree, options.projectName);
  addFiles(tree, options);
  await configureTailwindIntegration(tree, options);

  const e2eTask = await setupE2ETests(tree, options);
  if (e2eTask) {
    tasks.push(e2eTask);
  }

  const integrationsTask = addIntegrationsPackages(tree, options.integrations);
  if (integrationsTask) {
    tasks.push(integrationsTask);
  }

  await formatFiles(tree);

  return () => tasks.forEach((task) => task());
}

export default applicationGenerator;
