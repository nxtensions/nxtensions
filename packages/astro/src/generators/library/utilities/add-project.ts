import { addProjectConfiguration, readNxJson, type Tree } from '@nx/devkit';
import { addTargetDefaults } from '../../utilities/target-defaults';
import type { NormalizedGeneratorOptions } from '../schema';

export function addProject(
  tree: Tree,
  options: NormalizedGeneratorOptions
): void {
  const nxJson = readNxJson(tree);
  addTargetDefaults(tree, '@nxtensions/astro:check', {
    inputs:
      nxJson.namedInputs && 'production' in nxJson.namedInputs
        ? ['production', '^production']
        : ['default', '^default'],
    cache: true,
  });

  addProjectConfiguration(tree, options.projectName, {
    root: options.projectRoot,
    sourceRoot: options.projectRoot,
    projectType: 'library',
    targets: { check: { executor: '@nxtensions/astro:check' } },
    tags: options.tags,
  });
}
