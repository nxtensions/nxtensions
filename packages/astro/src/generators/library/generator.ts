import { formatFiles, GeneratorCallback, Tree } from '@nrwl/devkit';
import { initGenerator } from '../init/generator';
import { GeneratorOptions } from './schema';
import {
  addFiles,
  addPathMapping,
  addProject,
  normalizeOptions,
} from './utilities';

export async function libraryGenerator(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback> {
  if (rawOptions.publishable && !rawOptions.importPath) {
    throw new Error(
      'For publishable libraries the "--importPath" must be provided. Please note it needs to be a valid npm package name (e.g. my-lib or @my-org/my-lib).'
    );
  }

  const options = normalizeOptions(tree, rawOptions);

  const initTask = initGenerator(tree, { addCypressTests: false });

  addProject(tree, options);
  addFiles(tree, options);
  addPathMapping(tree, options);

  await formatFiles(tree);

  return initTask;
}

export default libraryGenerator;
