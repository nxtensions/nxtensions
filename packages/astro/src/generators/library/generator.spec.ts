jest.mock('@nx/devkit', () => ({
  ...jest.requireActual('@nx/devkit'),
  formatFiles: jest.fn(),
}));

import {
  formatFiles,
  readJson,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '../utilities/testing';
import { libraryGenerator } from './generator';
import { GeneratorOptions } from './schema';

describe('library generator', () => {
  let tree: Tree;
  const options: GeneratorOptions = { name: 'lib1' };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  test('should add project configuration', async () => {
    await libraryGenerator(tree, options);

    const config = readProjectConfiguration(tree, options.name);
    expect(config).toMatchSnapshot();
  });

  test('should generate files', async () => {
    await libraryGenerator(tree, options);

    expect(tree.exists(`${options.name}/src/index.js`)).toBeTruthy();
    expect(tree.exists(`${options.name}/src/lib/Lib1.astro`)).toBeTruthy();
    expect(tree.exists(`${options.name}/README.md`)).toBeTruthy();
    expect(tree.exists(`${options.name}/tsconfig.json`)).toBeTruthy();
  });

  test('should generate files in a monorepo layout', async () => {
    tree.write('libs/.gitkeep', '');

    await libraryGenerator(tree, options);

    expect(tree.exists(`libs/${options.name}/src/index.js`)).toBeTruthy();
    expect(tree.exists(`libs/${options.name}/src/lib/Lib1.astro`)).toBeTruthy();
    expect(tree.exists(`libs/${options.name}/README.md`)).toBeTruthy();
    expect(tree.exists(`libs/${options.name}/tsconfig.json`)).toBeTruthy();
  });

  test('should add the path mapping', async () => {
    await libraryGenerator(tree, options);

    const { compilerOptions } = readJson(tree, 'tsconfig.base.json');
    expect(compilerOptions.paths[`@proj/${options.name}`]).toStrictEqual([
      `${options.name}/src/index.js`,
    ]);
  });

  test('should format files', async () => {
    await libraryGenerator(tree, options);

    expect(formatFiles).toHaveBeenCalled();
  });

  describe('--directory', () => {
    const directory = 'some-directory/sub-directory';

    test('should add project with the right name when a directory is provided', async () => {
      await libraryGenerator(tree, { ...options, directory });

      const project = readProjectConfiguration(
        tree,
        `some-directory-sub-directory-${options.name}`
      );
      expect(project).toBeTruthy();
    });

    test('should generate files in the right directory', async () => {
      await libraryGenerator(tree, { ...options, directory });

      expect(
        tree.exists(`${directory}/${options.name}/src/index.js`)
      ).toBeTruthy();
      expect(
        tree.exists(`${directory}/${options.name}/src/lib/Lib1.astro`)
      ).toBeTruthy();
      expect(
        tree.exists(`${directory}/${options.name}/README.md`)
      ).toBeTruthy();
      expect(
        tree.exists(`${directory}/${options.name}/tsconfig.json`)
      ).toBeTruthy();
    });

    test('should generate files in the right directory in a monorepo layout', async () => {
      tree.write('libs/.gitkeep', '');

      await libraryGenerator(tree, { ...options, directory });

      expect(
        tree.exists(`libs/${directory}/${options.name}/src/index.js`)
      ).toBeTruthy();
      expect(
        tree.exists(`libs/${directory}/${options.name}/src/lib/Lib1.astro`)
      ).toBeTruthy();
      expect(
        tree.exists(`libs/${directory}/${options.name}/README.md`)
      ).toBeTruthy();
      expect(
        tree.exists(`libs/${directory}/${options.name}/tsconfig.json`)
      ).toBeTruthy();
    });

    test('should add the path mapping with the right directory', async () => {
      await libraryGenerator(tree, { ...options, directory });

      const { compilerOptions } = readJson(tree, 'tsconfig.base.json');
      expect(
        compilerOptions.paths[
          `@proj/some-directory-sub-directory-${options.name}`
        ]
      ).toStrictEqual([`${directory}/${options.name}/src/index.js`]);
    });
  });

  describe('--tags', () => {
    test('should add project tags when provided', async () => {
      await libraryGenerator(tree, { ...options, tags: 'foo, bar' });

      const { tags } = readProjectConfiguration(tree, options.name);
      expect(tags).toEqual(['foo', 'bar']);
    });
  });

  describe('--publishable', () => {
    test('should not generate a package.json when publishable is false', async () => {
      await libraryGenerator(tree, { ...options, publishable: false });

      expect(tree.exists(`${options.name}/package.json`)).toBeFalsy();
    });

    test('should throw when publishable is true and importPath was not specified', async () => {
      await expect(
        libraryGenerator(tree, { ...options, publishable: true })
      ).rejects.toThrow();
    });

    test('should generate a package.json when publishable is true', async () => {
      await libraryGenerator(tree, {
        ...options,
        publishable: true,
        importPath: 'lib1',
      });

      expect(tree.exists(`${options.name}/package.json`)).toBeTruthy();
    });
  });

  describe('--importPath', () => {
    const importPath = '@my-awesome-scope/lib1';

    test('should use the specified importPath as the package name', async () => {
      await libraryGenerator(tree, {
        ...options,
        importPath,
        publishable: true,
      });

      expect(readJson(tree, `${options.name}/package.json`).name).toBe(
        importPath
      );
    });

    test('should add the specified importPath to the path mappings in tsconfig.base.json', async () => {
      await libraryGenerator(tree, { ...options, importPath });

      const { compilerOptions } = readJson(tree, 'tsconfig.base.json');
      expect(compilerOptions.paths[importPath]).toStrictEqual([
        `${options.name}/src/index.js`,
      ]);
    });
  });
});
