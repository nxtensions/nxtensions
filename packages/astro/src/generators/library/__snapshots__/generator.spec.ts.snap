// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`library generator should add project configuration 1`] = `
{
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "name": "lib1",
  "projectType": "library",
  "root": "lib1",
  "sourceRoot": "lib1",
  "tags": [],
  "targets": {
    "check": {
      "executor": "@nxtensions/astro:check",
    },
  },
}
`;
